File: src/cn/yiiguxing/plugin/translate/ui/IconButton.java
package cn.yiiguxing.plugin.translate.ui;

import com.intellij.ui.components.panels.NonOpaquePanel;
import com.intellij.util.Consumer;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.awt.*;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;

/**
 * Icon Button.
 */
public class IconButton extends NonOpaquePanel {

    private final Icon icon;
    private final Icon pressIcon;
    private boolean isPressedByMouse;
    private boolean isActive = true;

    private final Consumer<MouseEvent> listener;

    public IconButton(@NotNull Icon icon, @Nullable Icon pressIcon, @NotNull Consumer<MouseEvent> listener) {
        this.icon = icon;
        this.pressIcon = pressIcon;
        this.listener = listener;

        addMouseListener(new MouseListener() {
            @Override
            public void mouseClicked(MouseEvent e) {
                IconButton.this.listener.consume(e);
            }

            @Override
            public void mousePressed(MouseEvent e) {
                isPressedByMouse = true;
                IconButton.this.repaint();
            }

            @Override
            public void mouseReleased(MouseEvent e) {
                isPressedByMouse = false;
                IconButton.this.repaint();
            }

            @Override
            public void mouseEntered(MouseEvent e) {
            }

            @Override
            public void mouseExited(MouseEvent e) {
                isPressedByMouse = false;
                IconButton.this.repaint();
            }
        });
    }

    @Override
    public Dimension getPreferredSize() {
        return new Dimension(icon.getIconWidth(), icon.getIconHeight());
    }

    public void setActive(final boolean active) {
        this.isActive = active;
        this.repaint();
    }

    @Override
    protected void paintComponent(Graphics g) {
        super.paintComponent(g);
        if (hasPaint()) {
            paintIcon(g, (!isActive || isPressedByMouse) && pressIcon != null ? pressIcon : icon);
        }
    }

    protected boolean hasPaint() {
        return getWidth() > 0;
    }

    private void paintIcon(@NotNull Graphics g, @NotNull Icon icon) {
        icon.paintIcon(this, g, 0, (getHeight() - icon.getIconHeight()) / 2);
    }

}

File: src/main/kotlin/cn/yiiguxing/plugin/translate/ui/IconButton.kt
package cn.yiiguxing.plugin.translate.ui

import com.intellij.ui.components.panels.NonOpaquePanel
import com.intellij.util.Consumer
import java.awt.Dimension
import java.awt.Graphics
import java.awt.event.MouseEvent
import java.awt.event.MouseListener
import javax.swing.Icon

/**
 * Icon Button.
 */
open class IconButton(
        private val icon: Icon,
        private val pressIcon: Icon?,
        private val listener: Consumer<MouseEvent>
) : NonOpaquePanel() {
    private var isPressedByMouse: Boolean = false
    private var isActive = true

    init {
        addMouseListener(object : MouseListener {
            override fun mouseClicked(e: MouseEvent) {
                this@IconButton.listener.consume(e)
            }

            override fun mousePressed(e: MouseEvent) {
                isPressedByMouse = true
                this@IconButton.repaint()
            }

            override fun mouseReleased(e: MouseEvent) {
                isPressedByMouse = false
                this@IconButton.repaint()
            }

            override fun mouseEntered(e: MouseEvent) {}

            override fun mouseExited(e: MouseEvent) {
                isPressedByMouse = false
                this@IconButton.repaint()
            }
        })
    }

    override final fun addMouseListener(l: MouseListener?) {
        super.addMouseListener(l)
    }

    override fun getPreferredSize(): Dimension {
        return Dimension(icon.iconWidth, icon.iconHeight)
    }

    fun setActive(active: Boolean) {
        this.isActive = active
        this.repaint()
    }

    override fun paintComponent(g: Graphics) {
        super.paintComponent(g)
        if (hasPaint()) {
            paintIcon(g, if ((!isActive || isPressedByMouse) && pressIcon != null) pressIcon else icon)
        }
    }

    protected open fun hasPaint() = width > 0

    protected open fun paintIcon(g: Graphics, icon: Icon) {
        icon.paintIcon(this, g, 0, (height - icon.iconHeight) / 2)
    }

}
